src/__experimental__/components/select-async/select-async.spec.js

```diff
@@ -0,0 +1,170 @@
+];
+const pageTwo = [
+  { displayed_as: 'Item 3', id: 3 },
+  { displayed_as: 'Item 4', id: 4 }
+];
+const validResponse = (items = pageOne, page = 1) => ({
+  data: {
+    $items: items,
+    $page: page,
+    $total: 4
+  }
+});
+const validRequest = ({ search, page, extra }) => ({
+  params: {
+    items_per_page: 20,
+    page,
+    search
+  },
+  ...extra
+});
+const mockResponse = (response) => {
+  axios.get = jest.fn((_, opts) => {
+    if (response) return response;
+    switch (opts.params.page) {
+      case 2:
+        return validResponse(pageTwo, 2);
+      default:
+        return validResponse(pageOne);
+    }
+  });
+};
+const openResults = (wrapper) => {
+  wrapper.find('input').simulate('focus');
+  return wrapper;
+};
+const findResults = wrapper => wrapper.update().find('Option');
+const getNextPage = wrapper => wrapper.find('Select').props().onLazyLoad();
+const filterResults = (wrapper, filter) => {
+  const select = wrapper.find('Select');
+  select.setState({ filter });
+  select.props().onFilter();
+};
+
+describe('SelectAsync', () => {
+  const render = (props, renderer = mount) => (
+    renderer(<SelectAsync endpoint={ endpoint } { ...props } />)
+  );
+
+  it('renders a Select component', () => {
+    expect(render(undefined, shallow)).toMatchSnapshot();
+  });
+
+  describe('on open', () => {
+    it('fetches data for the first page of results', async () => {
+      mockResponse();
+      const wrapper = await openResults(render());
+      expect(axios.get).toHaveBeenCalledWith(
+        endpoint, validRequest({ page: 1 })
+      );
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('no results', () => {
+    it('it renders with null', async () => {
+      mockResponse(validResponse([]));
+      const wrapper = await openResults(render());
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('custom children render', () => {
+    it('uses a custom method to render the children', async () => {
+      mockResponse();
+      const props = { children: items => items.map(() => <Option>custom!</Option>) };
+      const wrapper = await openResults(render(props));
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+  });
+
+  describe('custom request', () => {
+    it('allows developers to customise the request options', async () => {
+      mockResponse();
+      const props = { onRequest: () => ({ customRequest: true }) };
+      await openResults(render(props));
+      expect(axios.get).toHaveBeenCalledWith(
+        endpoint, { customRequest: true }
+      );
+    });
+  });
+
+  describe('custom response', () => {
+    it('throws an error if the response is not supported', async () => {
+      mockResponse({ data: {} });
+      try {
+        await render().instance().fetchData({ page: 1 });
+      } catch ({ message }) {
+        expect(message).toEqual(`The server response does not contain an $items attribute.
```

Comment:
Perhaps pop this in a different file and export it to both the test and component, so a change to error message doesn't break the test.

Reply:
alrighty

Reply:
✅